Introduce Node Lifecycle WG (#8396)
kubernetes/community: wg-node-lifecycle/charter.md

# WG Node Lifecycle Charter

This charter adheres to the conventions described in the [Kubernetes Charter README] and uses
the Roles and Organization Management outlined in [wg-governance].

[Kubernetes Charter README]: /committee-steering/governance/README.md

## Scope

The Kubernetes ecosystem currently faces challenges in node maintenance scenarios, with multiple

> projects independently addressing similar issues. The goal of this working group is to develop

Reviewer: Can we list out the references of the projects  if possible ?

Reviewer 2: Can we list out the references of the projects if possible ?

Author: This might be too heavy for the WG declaration. We have a partial list in https://github.com/atiratree/kube-enhancements/blob/improve-node-maintenance/keps/sig-apps/4212-declarative-node-maintenance/README.md#motivation and I expect other projects to be discussed/added in the future.

I would prefer to track it separately once the WG is established. Let's also see what others think. If there is a demand I can add it.

Reviewer 3: +1 to maybe listing this out. Or at least a link to where we can find more information

Reviewer 4: Here is one: https://github.com/Mellanox/maintenance-operator

Reviewer 5: We have built an internal version, some others are draino and medik8s

Author: Ok, let's add the list here as a starting point then.

Author: I have added a new `Relevant Projects` section.  Please let me know if you know of others, or want to somehow reference your internal ones.

unified APIs that the entire ecosystem can depend on, reducing the maintenance burden across

> projects and addressing scenarios that impede node drain or cause improper pod termination. Our

Reviewer 2: Is it the plan to add a unified API which control the kubernetes features in this area ? ( like Priority classes, PDBs..etc? ) 
If Pod scaling ..etc are in consideration , do we also need to consider Resource Quota and Auto scalers like HPA here?

  workloads. Also explore the interactions with other eviction mechanisms.
- Improve the Graceful/Non-Graceful Node Shutdown and consider how this affects the node lifecycle.
  To graduate the [Graceful Node Shutdown](https://github.com/kubernetes/enhancements/issues/2000)
  feature to GA and resolve the associated node shutdown issues.
- Improve the scheduling and pod/node autoscaling to take into account ongoing node maintenance and
  the new disruption model/evictions.
- Explore the cloud provider use cases and how they can hook in into the node lifecycle. So that the
  users can use the same APIs or configurations across the board. 
- Migrate users of the eviction based kubectl-like drain (kubectl, cluster autoscaler, karpenter,
  ...) to use the new approach.

### Out of scope

- Implementing cloud provider specific logic, the goal is to have high-level API that the providers
  can use, hook into, or extend.
- Infrastructure provisioning, deprovisioning solution or physical infrastructure lifecycle
  management solution.

## Stakeholders

- SIG Apps
- SIG Architecture
- SIG Autoscaling
- SIG CLI
- SIG Cloud Provider
- SIG Cluster Lifecycle

> - SIG Node

Reviewer 6: ```suggestion
- SIG Node
- SIG Network
```

  and other actors should use a new eviction API to gracefully remove pods. This should enable new
  migration strategies that prefer to surge (upscale) pods first rather than downscale them. It
  should also allow other users/components to monitor pods that are gracefully removed/terminated
  and provide better behaviour in terms of de/scheduling, scaling and availability. 
- As an end user, I cannot bear the cost of blue-green upgrades, especially with special hardware
  accelerators; it's far too expensive. It is more cost-effective to coordinate a drain and then
  upgrade.
- As a cloud provider, I need to perform regular maintenance on the hardware in my fleet. Enhancing
  Kubernetes to help CSPs safely remove hardware will reduce operational costs.
- Modelling the cost of doing accelerator maintenance in today's world can be massive. And since
  hardware accelerators tend to need more love and care, having software support to coordinate
  maintenance will reduce operational costs.

## Deliverables

The WG will coordinate requirement gatherthing and design, eventually leading to
KEP(s)s and code associated with the ideas.

> Area we expect to explore:

Reviewer 3: Node Lifecycle around accelerators always comes up. Is there consideration in this group to explore these areas?

Author: Yes, although I am not sure if there will be deliverables specifically targeting accelerators yet. We mention them in the user stories and will consider them when creating the APIs.

- An API to express node drain/maintenance.
  Currently tracked in https://github.com/kubernetes/enhancements/issues/4563.
- An API to solve the problems wrt the API-initiated Eviction API and PDBs.
  Currently tracked in https://github.com/kubernetes/enhancements/issues/4212.
- Introduce enhancements across multiple Kubernetes SIGs to add support for the new APIs to solve

>   wide range of issue.

Reviewer 6: Core APIs need a reference implementation that execute them and defined semantics with Conformance tests, otherwise this will become implementation specific and non-portable across distros

Author: Good idea to mention this explicitly - added below the list.

# WG Node Lifecycle Charter

This charter adheres to the conventions described in the [Kubernetes Charter README] and uses
the Roles and Organization Management outlined in [wg-governance].

[Kubernetes Charter README]: /committee-steering/governance/README.md

## Scope

The Kubernetes ecosystem currently faces challenges in node maintenance scenarios, with multiple
projects independently addressing similar issues. The goal of this working group is to develop
unified APIs that the entire ecosystem can depend on, reducing the maintenance burden across
projects and addressing scenarios that impede node drain or cause improper pod termination. Our

> objective is to create easily configurable, out-of-the-box solutions that seamlessly integrate with

Reviewer 7: So the goal is APIs to use by solutions or implement a solution? These two sentences seems to be at odds. Maybe mention that k8s has no plans blocking customers implementing advanced use cases

Author: We want to do both :) I have added another sentence there to explain it better.

existing APIs and behaviors.

To properly solve the node drain, we must first understand the node lifecycle. This includes
provisioning/sunsetting of the nodes, PodDisruptionBudgets, API-initiated eviction and node
shutdown. This then impacts both the node and pod autoscaling, load balancing, and the applications
running in the cluster. All of these areas have issues and would benefit from a unified approach.

> ### In scope

Reviewer 7: Another goal should include making Pods work reliably while terminating. This is important since with prolifiration of non-live migratable VMs with accelerators, we see more and more situations when maintanence-caused termination should be taking hours if not days.

Author: Good idea, I have added a two more stories. All in all, the _In scope_ section covers this in general, I hope.

> - Explore a unified way of draining the nodes and managing node maintenance by introducing new APIs

Reviewer 7: Can you please add a goal of migrating existing scenarios to the new API so the group will be tasked to not break users when they are upgrading

Author: We do have that in scope under

```
Migrate users of the eviction based kubectl-like drain (kubectl, cluster autoscaler, karpenter), and other scenarios to use the new approach.
```

So far it is pretty generic until we have a clearer vision. Please let me know if you would like to see something more specific.

  feature to GA and resolve the associated node shutdown issues.
- Improve the scheduling and pod/node autoscaling to take into account ongoing node maintenance and
  the new disruption model/evictions.
- Explore the cloud provider use cases and how they can hook in into the node lifecycle. So that the
  users can use the same APIs or configurations across the board. 
- Migrate users of the eviction based kubectl-like drain (kubectl, cluster autoscaler, karpenter,
  ...) to use the new approach.

### Out of scope

- Implementing cloud provider specific logic, the goal is to have high-level API that the providers
  can use, hook into, or extend.
- Infrastructure provisioning, deprovisioning solution or physical infrastructure lifecycle
  management solution.

## Stakeholders

- SIG Apps
- SIG Architecture
- SIG Autoscaling
- SIG CLI
- SIG Cloud Provider
- SIG Cluster Lifecycle
- SIG Network
- SIG Node
- SIG Scheduling
- SIG Storage

Stakeholders span from multiple SIGs to a broad set of end users,
public and private cloud providers, Kubernetes distribution providers,
and cloud provider end-users. Here are some user stories:

> - As a cluster admin I want to have a simple interface to initiate a node drain/maintenance without

Reviewer 7: Can we add a goal to explore the scenario of getting the historical perspective on why the node was terminated/drained/killed. This comes up very often and maybe we can help those scenarios in this WG. Various ideas like Node object "thumbstones" were discussed in the past.

Author: I am not really sure I fully understand this. I have added a new point to the _In scope_ section that mentions this. Feel free to write a GitHub suggestion.

  any required manual interventions. I also want to be able to observe the node drain via the API
  and check on its progress. I also want to be able to discover workloads that are blocking the node
  drain.
- To support the new features, node maintenance, scheduler, descheduler, pod autoscaling, kubelet
  and other actors should use a new eviction API to gracefully remove pods. This should enable new
  migration strategies that prefer to surge (upscale) pods first rather than downscale them. It
  should also allow other users/components to monitor pods that are gracefully removed/terminated
  and provide better behaviour in terms of de/scheduling, scaling and availability. 
- As an end user, I cannot bear the cost of blue-green upgrades, especially with special hardware
  accelerators; it's far too expensive. It is more cost-effective to coordinate a drain and then

>   upgrade.

Reviewer 7: this is a strong statement. Maybe we can say it more generically: Investigate the most cost effective ways to upgrade nodes with the expensive accelerators deployed on it.

Reviewer 8: The blue-green upgrade with accelerators use-case I think is an important one to mention in some way.  Drain and upgrade always have a cost - money, time, complexity, ect....  We're trying to say that the current ecosystem of APIs and Tools, or lack thereof, cause solutions to be more "expensive" than it should be.  We can rephrase to emphasize this.

Perhaps it would be better to focus on some specific examples showing the "cost".  E.g. some people want to keep specific accelerators they have been using because of ware and tear.

Reviewer 7: My concern mostly about wording. Even more cost effective is to force kill and recreate. Drain is not always the best path, depending on the workload

Author: Ack, we have changed the wording.

- As a cloud provider, I need to perform regular maintenance on the hardware in my fleet. Enhancing
  Kubernetes to help CSPs safely remove hardware will reduce operational costs.
- Modelling the cost of doing accelerator maintenance in today's world can be massive. And since
  hardware accelerators tend to need more love and care, having software support to coordinate
  maintenance will reduce operational costs.

## Deliverables

The WG will coordinate requirement gatherthing and design, eventually leading to
KEP(s)s and code associated with the ideas.

Area we expect to explore:

> - An API to express node drain/maintenance.

Reviewer 7: One problem I feel we will need to address is how to transition existing drain logic in various components to this new API. Having a new API without migrating old ways to this new API create "yet another" way to do it and requires end user to understand more draining logics

Author: There is still no upgrade path, since we have not even agreed on the solution. We don't want to break people using the current approaches. The main incentive to switch should be painless upgrades/maintenance and other benefits.

I expect that the main components/users that use the kubectl(-like) drain should not have a hard time using the new solution(s). However, I am not sure what it will look like for the GNS, for example.

> - An API to express node drain/maintenance.

Reviewer 7: why graceful termination that mentioned above is not listed here? Mosty curious

Author: Ah, sorry, I have fixed that. I am also open to other KEPs/documents that people would like to include here.

# WG Node Lifecycle Charter

This charter adheres to the conventions described in the [Kubernetes Charter README] and uses
the Roles and Organization Management outlined in [wg-governance].

[Kubernetes Charter README]: /committee-steering/governance/README.md

## Scope

The Kubernetes ecosystem currently faces challenges in node maintenance scenarios, with multiple
projects independently addressing similar issues. The goal of this working group is to develop
unified APIs that the entire ecosystem can depend on, reducing the maintenance burden across
projects and addressing scenarios that impede node drain or cause improper pod termination. Our
objective is to create easily configurable, out-of-the-box solutions that seamlessly integrate with
existing APIs and behaviors. We will strive to make these solutions minimalistic and extensible to
support advanced use cases across the ecosystem.

To properly solve the node drain, we must first understand the node lifecycle. This includes

> provisioning/sunsetting of the nodes, PodDisruptionBudgets, API-initiated eviction and node

Author: Scheduling is certainly an important part as well. I have added a mention of scheduling constraints to our goals.

shutdown. This then impacts both the node and pod autoscaling, load balancing, and the applications
running in the cluster. All of these areas have issues and would benefit from a unified approach.

### In scope

> - Explore a unified way of draining the nodes and managing node maintenance by introducing new APIs

Reviewer 5: Is it worth including something about DRA device taints/drains?

Author: Seems relevant to me as this affects the pod and device/node lifecycle. @pohly what do you think about including and discussing https://github.com/kubernetes/enhancements/issues/5055 in the WG?

Reviewer 9: Part of the motivation for 5055 is to *not* have to drain entire nodes when doing some maintenance which only affects one device or hardware managed by one driver - unless of course that maintenance ends up with having to reboot the node.

So I guess it depends?

Author: We would also like to handle the pod lifecycle better in any descheduling scenario (not just Node Drain/Maintenance). One option is to use the [EvictionRequest API](https://github.com/kubernetes/enhancements/issues/4563) which should give more power to the applications that are being disrupted.  So it might be interesting to see if we can make the disruption more graceful  in 5055.

Reviewer 5: > Part of the motivation for 5055 is to not have to drain entire nodes when doing some maintenance which only affects one device or hardware managed by one driver - unless of course that maintenance ends up with having to reboot the node.

The other end of this to consider though is devices that span multiple nodes

Author: I have added it to the WG. I think it would be good to discuss this feature and its impact . Also, it might be better to hold off beta for some time.

  the Node API to expose additional state or status in order to coalesce other core Kubernetes and
  community APIs around node lifecycle management.
- Improve the disruption model that is currently implemented by API-initiated Eviction API and PDBs.
  Improve the descheduling, availability and migration capabilities of today's application
  workloads. Also explore the interactions with other eviction mechanisms.
- Improve the Graceful/Non-Graceful Node Shutdown and consider how this affects the node lifecycle.
  To graduate the [Graceful Node Shutdown](https://github.com/kubernetes/enhancements/issues/2000)
  feature to GA and resolve the associated node shutdown issues.
- Improve the scheduling and pod/node autoscaling to take into account ongoing node maintenance and
  the new disruption model/evictions.
- Explore the cloud provider use cases and how they can hook in into the node lifecycle. So that the
  users can use the same APIs or configurations across the board.
- Migrate users of the eviction based kubectl-like drain (kubectl, cluster autoscaler, karpenter,
  ...) and other scenarios to use the new approach.

> - Explore possible scenarios behind the reason why the node was terminated/drained/killed and how to

Reviewer 5: It seems like everyone is solving the problem of node fault remediation independently and building private in house solutions. Improving the drain behavior is one aspect of remediation (generally the first step after detection) but this would still be good to collaborate on a community solution for remediation once a node is ready to be acted on as well. Thoughts on including this?

Author: Yeah, we want to include these considerations in the WG. We imply these in our goals, but I have added your suggestion as an additional user story to make it clearer.

  track and react to each of them. Consider past discussions/historical perspective
  (e.g. "thumbstones").

### Out of scope

- Implementing cloud provider specific logic, the goal is to have high-level API that the providers

>   can use, hook into, or extend.

Reviewer 10: +1

  hardware accelerators; it's far too expensive. I would like to choose a strategy on how to
  coordinate the node drain and the upgrade to achieve better cost-effectiveness.
- As a cloud provider, I need to perform regular maintenance on the hardware in my fleet. Enhancing
  Kubernetes to help CSPs safely remove hardware will reduce operational costs.
- Modelling the cost of doing accelerator maintenance in today's world can be massive. And since
  hardware accelerators tend to need more love and care, having software support to coordinate
  maintenance will reduce operational costs.
- As a user, I want to prevent any disruption to my pet or expensive workloads (VMs, ML with
  accelerators) and either prevent termination altogether or have a reliable migration path. 
  Features like `terminationGracePeriodSeconds` are not sufficient as the termination/migration can
  take hours if not days.
- As a user, I want my application to finish all network and storage operations before terminating a
  pod. This includes closing pod connections, removing pods from endpoints, writing cached writes
  to the underlying storage and completing storage cleanup routines.

> 

Reviewer 10: i think another user story is around the use of ephemeral low cost instances on cloud providers. eg

```
As a cluster admin, I would like to use a mixture of on-demand and temporary spot instances in  my clusters to reduce cloud expenditure. Having more reliable lifecycle and drain mechanisms for nodes will improve cluster stability in scenarios where instances may be terminated by the cloud provider due to cost-related thresholds.
```

Author: Agree, this story is also important to have. Thanks!

  coordinate the node drain and the upgrade to achieve better cost-effectiveness.
- As a cloud provider, I need to perform regular maintenance on the hardware in my fleet. Enhancing
  Kubernetes to help CSPs safely remove hardware will reduce operational costs.
- Modelling the cost of doing accelerator maintenance in today's world can be massive. And since
  hardware accelerators tend to need more love and care, having software support to coordinate
  maintenance will reduce operational costs.
- As a cluster admin, I would like to use a mixture of on-demand and temporary spot instances in my
  clusters to reduce cloud expenditure. Having more reliable lifecycle and drain mechanisms for
  nodes will improve cluster stability in scenarios where instances may be terminated by the cloud
  provider due to cost-related thresholds.
- As a user, I want to prevent any disruption to my pet or expensive workloads (VMs, ML with
  accelerators) and either prevent termination altogether or have a reliable migration path. 
  Features like `terminationGracePeriodSeconds` are not sufficient as the termination/migration can

>   take hours if not days.

Reviewer 5: Yeah I agree there's a problem that eviction API / drain doesn't guarantee it will finish within a reasonable time, especially if the node is having issues (things get stuck terminating etc.). 

But this at least we can do today right?

> an accelerator is degraded on a Node, so I don't want to run future workloads there but it's ok for the current one to finish

You can just taint the node or the devices with NoSchedule.

>   take hours if not days.

Reviewer 8: >You can just taint the node or the devices with NoSchedule.
>
Yes, that is a solution assuming that:
1) workloads will eventually drain if we wait long enough
2) all termination steps will be successful

However, 1) can theoretically always work but it is the slowest possible solution and 2) is not guaranteed to work.

>   take hours if not days.

Author: > I'm aware of k8s-shredder, an Adobe project that puts nodes into maintenance and then gives them a week to be clear before removing them. I'm wondering if this case is to say, even in that scenario, don't kill my workload?

This is a good example. We want the applications/admins be aware of upcoming maintenances. But also pods in most of descheduling scenarios so that they are given opportunity to migrate or cleanup before the termination, which is hard to do with PDBs.

The goal is not to override PDBs (it is also hard to do without breaking someone), the goal is to have a smarter layer above the PDBs.

>   take hours if not days.

Reviewer 2: If `eviction` API has been used and if the `terminationGracePeriod` passed, isnt there a force kill executed ? Is there a scenario where the pod get stuck for days/weeks in this case?

>   take hours if not days.

Author: Just the eviction API alone can cause pods to get stuck. All in all, I would prefer we do not dive deep into the topic and focus mostly on the scope in this PR.

>   take hours if not days.

Reviewer 2: Hmm.. ooc, can you please share any reference issues where `eviction API` itself get stuck? while I understand the scope to an extent of this WG, I feel , the need to have a new WG itself arised due to the lack of co-orindation of the Primitive/features we have in K/K around scheduling, Preemption and Eviction.. imo, with the new design if we still not address half of the issues on this area, it dont serve the purpose.

>   take hours if not days.

Author: When the eviction gets through; i.e. results in a delete call then everything works fine. The stuck eviction can be caused by:
- PDB itself: that is the pods don't satisfy the constraints (and never will)
- catching the eviction request via a validation webhook and discarding it (e.g. some projects from https://github.com/atiratree/kube-enhancements/blob/improve-node-maintenance/keps/sig-apps/4212-declarative-node-maintenance/README.md#motivation use it)

> co-ordination or complexity of the Primitive/features we have in K/K around scheduling, Preemption and Eviction.. imo, with the new design if we still not address half of the issues on this area, it dont serve the purpose.

Yes, this is one of the important main goals that should be solved even before we introduce a node maintenance API IMO. Added to the goals to make it clear.

- As a user, I want my application to finish all network and storage operations before terminating a
  pod. This includes closing pod connections, removing pods from endpoints, writing cached writes

>   to the underlying storage and completing storage cleanup routines.

Reviewer 5: The way we handle this is our internal drain API has a label selector for things to ignore and by default just ignores daemonsets instead of worrying about the ordering here. Daemonsets aren't really supported under drain anyway. That should handle most cases for system level things like csi/cni/etc.

>   to the underlying storage and completing storage cleanup routines.

Author: Yes, daemonset pods should be considered as part of the maintenance scenarios, especially in cases when the node is going to shutdown. I have added it to goals to make it clear.

We also had discussions with SIG node about static pod termination in the past, and they were generally not against it. But we lack use cases for it so far.

>   to the underlying storage and completing storage cleanup routines.

Reviewer 2: iiuc, this proposal aim to define an order for static pod termination, DaemonSet pods or other `system-node-critical` Priority class Pod termination in node drain scenario.  Is that a correct assumption ?

>   to the underlying storage and completing storage cleanup routines.

Author: I am not saying that we will solve static pod termination, just that we will look into that :) And yes, I think there should definitely be an ordering for both DS/Static.

  community APIs around node lifecycle management.
- Improve the disruption model that is currently implemented by API-initiated Eviction API and PDBs.
  Improve the descheduling, availability and migration capabilities of today's application
  workloads. Also explore the interactions with other eviction mechanisms.
- Improve the Graceful/Non-Graceful Node Shutdown and consider how this affects the node lifecycle.
  To graduate the [Graceful Node Shutdown](https://github.com/kubernetes/enhancements/issues/2000)
  feature to GA and resolve the associated node shutdown issues.
- Improve the scheduling and pod/node autoscaling to take into account ongoing node maintenance and
  the new disruption model/evictions. This includes balancing of the pods according to scheduling
  constraints. 
- Consider improving the pod lifecycle of DaemonSets and Static pods during a node maintenance.
- Explore the cloud provider use cases and how they can hook in into the node lifecycle. So that the
  users can use the same APIs or configurations across the board.
- Migrate users of the eviction based kubectl-like drain (kubectl, cluster autoscaler, karpenter,

>   ...) and other scenarios to use the new approach.

Reviewer 11: nit: Make it more clear that when you say "new approach" you're referring to "a unified way of draining the nodes" you mentioned in the first bullet point. Either move this sentence up or clarify it here.

  track and react to each of them. Consider past discussions/historical perspective
  (e.g. "thumbstones").

### Out of scope

- Implementing cloud provider specific logic, the goal is to have high-level API that the providers
  can use, hook into, or extend.
- Infrastructure provisioning, deprovisioning solution or physical infrastructure lifecycle
  management solution.

## Stakeholders

- SIG Apps
- SIG Architecture
- SIG Autoscaling
- SIG CLI
- SIG Cloud Provider
- SIG Cluster Lifecycle
- SIG Network
- SIG Node
- SIG Scheduling
- SIG Storage

Stakeholders span from multiple SIGs to a broad set of end users,
public and private cloud providers, Kubernetes distribution providers,
and cloud provider end-users. Here are some user stories:

- As a cluster admin I want to have a simple interface to initiate a node drain/maintenance without
  any required manual interventions. I also want to be able to observe the node drain via the API
  and check on its progress. I also want to be able to discover workloads that are blocking the node
  drain.
- To support the new features, node maintenance, scheduler, descheduler, pod autoscaling, kubelet

>   and other actors should use a new eviction API to gracefully remove pods. This should enable new

Reviewer 11: nit: change "should" to something like "I want to" given that the KEP hasn't been accepted yet

ecosystem. This should be accompanied by E2E / Conformance tests.

## Relevant Projects

This is a list of known projects that solve similar problems in the ecosystem or would benefit from
the efforts of this WG:

- https://github.com/aws/aws-node-termination-handler
- https://github.com/foriequal0/pod-graceful-drain
- https://github.com/kubereboot/kured
- https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler
- https://github.com/kubernetes-sigs/karpenter
- https://github.com/kubevirt/kubevirt
- https://github.com/medik8s/node-maintenance-operator
- https://github.com/Mellanox/maintenance-operator
- https://github.com/openshift/machine-config-operator
- https://github.com/planetlabs/draino
- https://github.com/strimzi/drain-cleaner

There are also internal custom solutions that companies use.

## Roles and Organization Management

This WG adheres to the Roles and Organization Management outlined in [wg-governance]
and opts-in to updates and modifications to [wg-governance].

[wg-governance]: /committee-steering/governance/wg-governance.md

## Timelines and Disbanding

> The working group will disband when the KEPs we create are completed. We will

Reviewer 11: ```suggestion
The working group will disband once the core APIs defined in the KEPs have reached a stable state (GA) and ongoing maintenance ownership is established within the relevant SIGs. We will
```

Author: Thanks for the suggestions! Hope it is clear now.

- An API to express node drain/maintenance.
  Currently tracked in https://github.com/kubernetes/enhancements/issues/4212.
- An API to solve the problems wrt the API-initiated Eviction API and PDBs.
  Currently tracked in https://github.com/kubernetes/enhancements/issues/4563.
- An API/mechanism to gracefully terminate pods during a node shutdown.
  Graceful node shutdown feature tracked in https://github.com/kubernetes/enhancements/issues/2000.
- An API to deschedule pods that use DRA devices.
  DRA: device taints and tolerations feature tracked in https://github.com/kubernetes/enhancements/issues/5055.
- An API to remove pods from endpoints before they terminate.
  Currently tracked in https://docs.google.com/document/d/1t25jgO_-LRHhjRXf4KJ5xY_t8BZYdapv7MDAxVGY6R8/edit?tab=t.0#heading=h.i4lwa7rdng7y.
- Introduce enhancements across multiple Kubernetes SIGs to add support for the new APIs to solve
  wide range of issue.

We expect to provide reference implementations of the new APIs including but not limited to
controllers, API validation, integration with existing core components and extension points for the
ecosystem. This should be accompanied by E2E / Conformance tests.

> ## Relevant Projects

Reviewer 12: Here is a possible addition.
https://github.com/pfnet-research/node-operation-controller

My company developed and uses this project to automate node operation (e.g. taint and reboot) according to node conditions.
It also has NodeDisruptionBudget to limit the number of operations running concurrently to avoid e.g. control plane disruption.

# WG Node Lifecycle Charter

This charter adheres to the conventions described in the [Kubernetes Charter README] and uses
the Roles and Organization Management outlined in [wg-governance].

[Kubernetes Charter README]: /committee-steering/governance/README.md

## Scope

The Kubernetes ecosystem currently faces challenges in node maintenance scenarios, with multiple
projects independently addressing similar issues. The goal of this working group is to develop
unified APIs that the entire ecosystem can depend on, reducing the maintenance burden across
projects and addressing scenarios that impede node drain or cause improper pod termination. Our
objective is to create easily configurable, out-of-the-box solutions that seamlessly integrate with
existing APIs and behaviors. We will strive to make these solutions minimalistic and extensible to

> support advanced use cases across the ecosystem.

Author: Ack, the goals should be stated in this fashion.

the applications running in the cluster. All of these areas have issues and would benefit from a
unified approach.

### In scope

- Explore a unified way of draining the nodes and managing node maintenance by introducing new APIs
  and extending the current ones. This includes exploring extension to or interactions with the Node
  object.
- Analyze the node lifecycle, the Node API, and possible interactions. We want to explore augmenting
  the Node API to expose additional state or status in order to coalesce other core Kubernetes and
  community APIs around node lifecycle management.
- Improve the disruption model that is currently implemented by API-initiated Eviction API and PDBs.
  Improve the descheduling, availability and migration capabilities of today's application
  workloads. Also explore the interactions with other eviction mechanisms.
- Coordinate pod termination and issues around de/scheduling, preemption and eviction.
- Improve the Graceful/Non-Graceful Node Shutdown and consider how this affects the node lifecycle.
  To graduate the [Graceful Node Shutdown](https://github.com/kubernetes/enhancements/issues/2000)

>   feature to GA and resolve the associated node shutdown issues.

Author: This was requested by SIG Node. @SergeyKanzhelev can you please give input how would you like the goals to be defined?

>   feature to GA and resolve the associated node shutdown issues.

Author: I have removed the link and simplified this point.

- Improve the scheduling and pod/node autoscaling to take into account ongoing node maintenance and
  the new disruption model/evictions. This includes balancing of the pods according to scheduling
  constraints. 
- Consider improving the pod lifecycle of DaemonSets and Static pods during a node maintenance.

> - Explore the cloud provider use cases and how they can hook in into the node lifecycle. So that the

Reviewer 10: minor nit

```suggestion
- Explore the cloud provider use cases and how they can hook into the node lifecycle. So that the
```

- Explore possible scenarios behind the reason why the node was terminated/drained/killed and how to
  track and react to each of them. Consider past discussions/historical perspective
  (e.g. "thumbstones").

### Out of scope

- Implementing cloud provider specific logic, the goal is to have high-level API that the providers
  can use, hook into, or extend.
- Infrastructure provisioning, deprovisioning solution or physical infrastructure lifecycle
  management solution.

## Stakeholders

- SIG Apps
- SIG Architecture
- SIG Autoscaling
- SIG CLI
- SIG Cloud Provider
- SIG Cluster Lifecycle
- SIG Network
- SIG Node
- SIG Scheduling
- SIG Storage

Stakeholders span from multiple SIGs to a broad set of end users,
public and private cloud providers, Kubernetes distribution providers,
and cloud provider end-users. Here are some user stories:

- As a cluster admin I want to have a simple interface to initiate a node drain/maintenance without
  any required manual interventions. I also want to be able to observe the node drain via the API
  and check on its progress. I also want to be able to discover workloads that are blocking the node

>   drain.

Reviewer 13: Nit: this entire section has 3 separate use-cases:
1. initiate
2. observe
3. discover 

Can you just split them accordingly. It's easier to read shorter user stories.

Author: Ack, I will go over the use cases and improve them.

Author: I have split this story into two; one for the initation and one for the observability.

- To support the new features, node maintenance, scheduler, descheduler, pod autoscaling, kubelet,
  and other actors want to use a new eviction API to gracefully remove pods. This would enable new
  migration strategies that prefer to surge (upscale) pods first rather than downscale them. It
  would also allow other users/components to monitor pods that are gracefully removed/terminated
  and provide better behaviour in terms of de/scheduling, scaling and availability.

> - As a cluster admin, I want to be able to perform arbitrary actions after the node drain is

Reviewer 13: This one I'd move up with the other cases where you're describing admin perspective.

Author: Done, also moved some of the other user stories.

  complete, such as resetting GPU drivers, resetting NICs, performing software updates or shutting
  down the machine.
- As an end user, I would like more alternatives to blue-green upgrades, especially with special
  hardware accelerators; it's far too expensive. I would like to choose a strategy on how to
  coordinate the node drain and the upgrade to achieve better cost-effectiveness.
- As a cloud provider, I need to perform regular maintenance on the hardware in my fleet. Enhancing
  Kubernetes to help CSPs safely remove hardware will reduce operational costs.

> - Modelling the cost of doing accelerator maintenance in today's world can be massive. And since

Reviewer 13: I'm having a hard time understanding this use case. I mean I know what you're trying to say but it's a bit convoluted. Can you re-phrase it to sound more like a use-case?

Author: @rthallisey can you please take a look?

Reviewer 8: Sure.  See if this is any better:

`As a cluster admin, I want to reduce the cost of doing maintenance on my hardware accelerators by using control-plane APIs to help coordinate maintenance and drain a Node.`

Author: Thank you, I have updated this user story.

  hardware accelerators tend to need more love and care, having software support to coordinate
  maintenance will reduce operational costs.

> - As a cluster admin, I would like to use a mixture of on-demand and temporary spot instances in my

Reviewer 13: This particular use-case is in contrast with the `Out of scope`. The idea of spot instances and in general any kind of instances shouldn't be part of the overall node lifecycle.

Author: I do not expect that we will explicitly support spot instances, but we may need to evaluate scenarios to support their node drain / pod termination flow.

  clusters to reduce cloud expenditure. Having more reliable lifecycle and drain mechanisms for
  nodes will improve cluster stability in scenarios where instances may be terminated by the cloud
  provider due to cost-related thresholds.

> - As a user, I want to prevent any disruption to my pet or expensive workloads (VMs, ML with

Author: I agree. This wording is too strong. I have replaced `prevent termination altogether` with `prevent termination for a long period of time`

  accelerators) and either prevent termination altogether or have a reliable migration path. 
  Features like `terminationGracePeriodSeconds` are not sufficient as the termination/migration can
  take hours if not days.

> - As a user, I want my application to finish all network and storage operations before terminating a

Reviewer 13: This particular use-case seems identical to the previous one. So I'd just drop it.

Author: There is a small distinction here. This point also applies to cheap/cattle applications. I think it is useful to keep it separate as we have a couple of issues/proposals for this.

  pod. This includes closing pod connections, removing pods from endpoints, writing cached writes
  to the underlying storage and completing storage cleanup routines.

## Deliverables

The WG will coordinate requirement gathering and design, eventually leading to
KEP(s)s and code associated with the ideas.

Area we expect to explore:

- An API to express node drain/maintenance.

>   Currently tracked in https://github.com/kubernetes/enhancements/issues/4212.

Reviewer 13: As mentioned elsewhere, drop these links. I'd focus on goals, not actual KEPs in charter. KEPs can and should be tracked as part of the ongoing work of a particular WG.

Author: Ack, removed, but added them to the `Relevant KEPs and Documents` section before we have an official place where to track them.

- An API to solve the problems wrt the API-initiated Eviction API and PDBs.
  Currently tracked in https://github.com/kubernetes/enhancements/issues/4563.
- An API/mechanism to gracefully terminate pods during a node shutdown.
  Graceful node shutdown feature tracked in https://github.com/kubernetes/enhancements/issues/2000.

> - An API to deschedule pods that use DRA devices.

Reviewer 13: Are you saying there will be separate API for descheduling any Pod and a Pod with DRA device? Why both can't just use `/evict`?

Author: This just references an existing feature without specifying the implementation details here.

  DRA: device taints and tolerations feature tracked in https://github.com/kubernetes/enhancements/issues/5055.

> - An API to remove pods from endpoints before they terminate.

Author: Also this is referencing an existing doc and I believe the `/evict` API is not sufficient in this scenario since it needs to apply to all workloads.

  Currently tracked in https://docs.google.com/document/d/1t25jgO_-LRHhjRXf4KJ5xY_t8BZYdapv7MDAxVGY6R8/edit?tab=t.0#heading=h.i4lwa7rdng7y.

> - Introduce enhancements across multiple Kubernetes SIGs to add support for the new APIs to solve

Author: We still do not know how the integration will look like, but it will have to be explored and can result in other enhancements. I would still prefer if we could reference such future work.

- Coordinate pod termination and issues around de/scheduling, preemption and eviction.
- Improve the Graceful/Non-Graceful Node Shutdown and consider how this affects the node lifecycle.
  To graduate the [Graceful Node Shutdown](https://github.com/kubernetes/enhancements/issues/2000)
  feature to GA and resolve the associated node shutdown issues.
- Improve the scheduling and pod/node autoscaling to take into account ongoing node maintenance and
  the new disruption model/evictions. This includes balancing of the pods according to scheduling
  constraints. 
- Consider improving the pod lifecycle of DaemonSets and static pods during a node maintenance.
- Explore the cloud provider use cases and how they can hook into the node lifecycle. So that the
  users can use the same APIs or configurations across the board.
- Migrate users of the eviction based kubectl-like drain (kubectl, cluster autoscaler, karpenter,
  ...) and other scenarios to use the new unified node draining approach.
- Explore possible scenarios behind the reason why the node was terminated/drained/killed and how to
  track and react to each of them. Consider past discussions/historical perspective

>   (e.g. "tombstones").

Reviewer 14: In SIG node this week, we saw a presentation on [node readiness](https://github.com/kubernetes/kubernetes/issues/131208) from @ajaysundark, and we have also been discussing the idea of [node capabilities](https://github.com/kubernetes-sigs/scheduler-plugins/pull/812#issuecomment-2689201925), which @pravk03 (among others) is looking into. These seem like related but slightly different topics from each other as well as from the existing taints/tolerations functionality. Can we include these in the scope of this WG?

Reviewer 15: +1 to adding Node Readiness in one of the goals. I can only speak from storage perspective, but one of the long standing problem is to make sure we can only schedule pods that require CSI driver, when said CSI driver is available on the node. 

Currently, we don't take this into account and hence a node can have too many pods that require storage scheduled to a node and they remain stuck and they require manual intervention to cleanup.I have added my thoughts here - https://github.com/kubernetes/kubernetes/issues/131208#issuecomment-2822334626

Reviewer 16: I agree that we should consider node readiness here, and I also want to call out the risk of scope creep of this WG. In the beginning I think we should be very clear about what our definition of done is. this proposal began as a way to close GNS KEP, and while I agree the expanded scope of reworking node API semantics would properly fix the problem, I want to be wary of biting off more than the community can collectively chew.

Reviewer 17: Thanks for including [node-readiness](https://github.com/kubernetes/enhancements/issues/5233) into consideration for wg-node-lifecycle.
I'm having the proposal ready for feedbacks http://docs/document/d/11i2_rewvcbQkFFq1BIwHa7lgIefgZ8Mak-QMZjP7fFs?tab=t.0. Happy to collaborate to align on the directions. Please include me relevant channels / discussions.

Author: Cool, I will take a look and we can add it to the WG agenda.

Btw, the link is wrong above but correct in the enhancements issue.

## Relevant KEPs and Documents

- https://github.com/kubernetes/enhancements/issues/4212
- https://github.com/kubernetes/enhancements/issues/4563
- https://github.com/kubernetes/enhancements/issues/2000
- https://github.com/kubernetes/enhancements/issues/5055
- https://docs.google.com/document/d/1t25jgO_-LRHhjRXf4KJ5xY_t8BZYdapv7MDAxVGY6R8/edit?tab=t.0#heading=h.i4lwa7rdng7y

## Relevant Projects

This is a list of known projects that solve similar problems in the ecosystem or would benefit from
the efforts of this WG:

- https://github.com/aws/aws-node-termination-handler
- https://github.com/foriequal0/pod-graceful-drain
- https://github.com/jukie/karpenter-deprovision-controller
- https://github.com/kubereboot/kured
- https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler
- https://github.com/kubernetes-sigs/cluster-api/
- https://github.com/kubernetes-sigs/karpenter
- https://github.com/kubernetes-sigs/kubespray
- https://github.com/kubevirt/kubevirt
- https://github.com/medik8s/node-maintenance-operator
- https://github.com/Mellanox/maintenance-operator
- https://github.com/openshift/machine-config-operator
- https://github.com/planetlabs/draino

> - https://github.com/strimzi/drain-cleaner

Author: +1, added

- As a user, I want my application to finish all network and storage operations before terminating a
  pod. This includes closing pod connections, removing pods from endpoints, writing cached writes
  to the underlying storage and completing storage cleanup routines.

## Deliverables

The WG will coordinate requirement gathering and design, eventually leading to implementation.

Area we expect to explore:

- An API to express node drain/maintenance.
- An API to solve the problems wrt the API-initiated Eviction API and PDBs.
- An API/mechanism to gracefully terminate pods during a node shutdown.
- An API to deschedule pods that use DRA devices.
- An API to remove pods from endpoints before they terminate.
- An API to track the schedulability (e.g. readiness) and capabilities of the node.
- Introduce enhancements across multiple Kubernetes SIGs to add support and integration for the new
  APIs to solve wide range of issues.

We expect to provide reference implementations of the new APIs including but not limited to
controllers, API validation, integration with existing core components and extension points for the
ecosystem. This should be accompanied by E2E / Conformance tests.

> ## Relevant KEPs and Documents

Reviewer 18: How about "Node Resource Hot Plug"? https://github.com/kubernetes/enhancements/issues/3953

Author: Ack, good to discuss. Although I am not sure how much intersection there is.

Reviewer 19: Thanks for suggesting and adding it here.

  coordinate the node drain and the upgrade to achieve better cost-effectiveness.
- As a cloud provider, I need to perform regular maintenance on the hardware in my fleet. Enhancing
  Kubernetes to help CSPs safely remove hardware will reduce operational costs.
- As an end user or admin, I would like to use a mixture of on-demand and temporary spot instances
  in my clusters to reduce cloud expenditure. Having more reliable lifecycle and drain mechanisms
  for nodes will improve cluster stability in scenarios where instances may be terminated by the
  cloud provider due to cost-related thresholds.
- As a user, I want to prevent any disruption to my pet or expensive workloads (VMs, ML with
  accelerators) and either prevent termination for a long period of time or have a reliable
  migration path. Features like `terminationGracePeriodSeconds` are not sufficient as the
  termination/migration can take hours if not days.
- As a user, I want my application to finish all network and storage operations before terminating a
  pod. This includes closing pod connections, removing pods from endpoints, writing cached writes
  to the underlying storage and completing storage cleanup routines.

> 

Reviewer 20: I suggest a new story to cover storage use case:

* As a cluster admin, I want a node to be declared as fully drained after all volumes are unmounted from it.

We can extend it to any "cleanup tasks" needed by other parts of kubelet (DRA?).

Today, a node is considered drained only when Pods are terminated. But storage cleanup is still in progress. A node can be rebooted / destroyed with still mounted volumes. This can lead to different states in the storage backend (who thinks the volume is still mounted) and OS + kubelet on the rebooted node (who thinks it's not mounted).

Reviewer 20: Original issue: https://github.com/kubernetes/kubernetes/issues/116618 / section about node drain

Author: I have added the story there. Thanks for the details!

  Kubernetes to help CSPs safely remove hardware will reduce operational costs.
- As an end user or admin, I would like to use a mixture of on-demand and temporary spot instances
  in my clusters to reduce cloud expenditure. Having more reliable lifecycle and drain mechanisms
  for nodes will improve cluster stability in scenarios where instances may be terminated by the
  cloud provider due to cost-related thresholds.
- As a user, I want to prevent any disruption to my pet or expensive workloads (VMs, ML with
  accelerators) and either prevent termination for a long period of time or have a reliable
  migration path. Features like `terminationGracePeriodSeconds` are not sufficient as the
  termination/migration can take hours if not days.
- As a user, I want my application to finish all network and storage operations before terminating a
  pod. This includes closing pod connections, removing pods from endpoints, writing cached writes
  to the underlying storage and completing storage cleanup routines.
- As a cluster admin, I want a node to be declared as fully drained after all volumes are unmounted
  from it.

## Deliverables

> The WG will coordinate requirement gathering and design, eventually leading to implementation.

Reviewer 7: ```suggestion
The WG will coordinate requirement gathering, design, implementation, and progressing thru graduation stages.
```

> Area we expect to explore:

Reviewer 7: ```suggestion

The group will help coordinate existing KEPs graduation as well as exploring new APIs and scenarios.

Area we expect to explore:
```

Author: Sounds good, thx!

- Node Resource Hot Plug: https://github.com/kubernetes/enhancements/issues/3953

## Relevant Projects

This is a list of known projects that solve similar problems in the ecosystem or would benefit from
the efforts of this WG:

- https://github.com/aws/aws-node-termination-handler
- https://github.com/foriequal0/pod-graceful-drain
- https://github.com/jukie/karpenter-deprovision-controller
- https://github.com/kubereboot/kured
- https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler
- https://github.com/kubernetes-sigs/cluster-api/
- https://github.com/kubernetes-sigs/karpenter
- https://github.com/kubernetes-sigs/kubespray
- https://github.com/kubevirt/kubevirt
- https://github.com/medik8s/node-maintenance-operator
- https://github.com/Mellanox/maintenance-operator
- https://github.com/NVIDIA/pika
- https://github.com/openshift/machine-config-operator
- https://github.com/planetlabs/draino
- https://github.com/strimzi/drain-cleaner

There are also internal custom solutions that companies use.

## Prioritization

> The group activity will focus on the following features:

Reviewer 21: Agree with these three, do we want to call out that we need to get consensus on EvictionRequest as a dependency of Declarative Node Maintenance?

projects independently addressing similar issues. The goal of this working group is to develop
unified APIs that the entire ecosystem can depend on, reducing the maintenance burden across
projects and addressing scenarios that impede node drain or cause improper pod termination. Our
objective is to create easily configurable, out-of-the-box solutions that seamlessly integrate with
existing APIs and behaviors. We will strive to make these solutions minimalistic and extensible to
support advanced use cases across the ecosystem.

To properly solve the node drain, we must first understand the node lifecycle. This includes
provisioning/sunsetting of the nodes, PodDisruptionBudgets, API-initiated eviction and node
shutdown. This then impacts both the node and pod autoscaling, de/scheduling, load balancing, and
the applications running in the cluster. All of these areas have issues and would benefit from a
unified approach.

### In scope

- Explore a unified way of draining the nodes and managing node maintenance by introducing new APIs
  and extending the current ones. This includes exploring extension to or interactions with the Node
  object.

> - Analyze the node lifecycle, the Node API, and possible interactions. We want to explore augmenting

Reviewer 22: I think we could have been more specific here. For example, analyse the possibility to set new conditions onto nodes.

  the Node API to expose additional state or status in order to coalesce other core Kubernetes and
  community APIs around node lifecycle management.
- Improve the disruption model that is currently implemented by API-initiated Eviction API and PDBs.
  Improve the descheduling, availability and migration capabilities of today's application
  workloads. Also explore the interactions with other eviction mechanisms.
- Coordinate pod termination and issues around, de/scheduling, preemption, eviction and readiness
  probes.
- Improve the Graceful/Non-Graceful Node Shutdown and consider how this affects the node lifecycle.
- Improve the scheduling and pod/node autoscaling to take into account ongoing node maintenance and
  the new disruption model/evictions. This includes balancing of the pods according to scheduling
  constraints. 
- Consider improving the pod lifecycle of DaemonSets and static pods during a node maintenance.
- Explore the cloud provider use cases and how they can hook into the node lifecycle. So that the
  users can use the same APIs or configurations across the board.

> - Migrate users of the eviction based kubectl-like drain (kubectl, cluster autoscaler, karpenter,

Reviewer 22: I am a bit sad that kured was removed here, while it was in the initial comments on a previous issue.

I would like to adapt kured to this framework at least.

  application in such scenarios.

## Deliverables

The WG will coordinate requirement gathering, design, implementation, and progressing through
graduation stages.

The group will help coordinate existing Kubernetes Enhancement Proposals (KEPs) graduation as well
as exploring new APIs and scenarios.

Area we expect to explore:

- An API to express node drain/maintenance.
- An API to solve the problems with the API-initiated Eviction API and PDBs.
- An API/mechanism to gracefully terminate pods during a node shutdown.
- An API to deschedule pods that use DRA devices.
- An API to remove pods from endpoints before they terminate.
- An API to track the schedulability (e.g. readiness) and capabilities of the node.
- Introduce enhancements across multiple Kubernetes SIGs to add support and integration for the new
  APIs to solve wide range of issues.

We expect to provide reference implementations of the new APIs including but not limited to
controllers (kube-controller-manager), API validation, integration with existing core components and
extension points for the ecosystem. This should be accompanied by E2E / Conformance tests.

> 

Reviewer 22: And cli...

- Improve the Graceful/Non-Graceful Node Shutdown and consider how this affects the node lifecycle.
- Improve the scheduling and pod/node autoscaling to take into account ongoing node maintenance and
  the new disruption model/evictions. This includes balancing of the pods according to scheduling
  constraints. 
- Consider improving the pod lifecycle of DaemonSets and static pods during a node maintenance.
- Explore the cloud provider use cases and how they can hook into the node lifecycle. So that the
  users can use the same APIs or configurations across the board.
- Migrate users of the eviction based kubectl-like drain (kubectl, cluster autoscaler, karpenter,
  ...) and other scenarios to use the new unified node draining approach.
- Explore possible scenarios behind the reason why the node was terminated/drained/killed and how to
  track and react to each of them. Consider past discussions/historical perspective
  (e.g. "tombstones").
- Explore feedback mechanism for ensuring schedulability (e.g. readiness) and capabilities of the 
  node. These can apply during the provisioning of the node, but also during the rest of the
  node lifecycle.

### Out of scope

> - Implementing cloud provider specific logic, the goal is to have high-level API that the providers

Reviewer 6: @atiratree it is important to. mention that Core APIs that are within kubernetes/kubernetes repo MUST have a reference implementation that allow to test these APIs to guarantee the semantics across the ecosystem.

The alternative is to follow a model like Gateway API where official CRDs with API review are released and follow an independent conformance process https://github.com/kubernetes-sigs/gateway-api

What will be the model of development of this high level APIs? can we be explicit about it?

  any required manual interventions.
- As a cluster admin, I want to be able to observe the node drain via the API and check on its
  progress. I also want to be able to discover workloads that are blocking the node
  drain.
- As a cluster admin, I want to be able to perform arbitrary actions after the node drain is
  complete, such as resetting GPU drivers, resetting NICs, performing software updates or shutting
  down the machine.
- As a cluster admin, I want to reduce the cost of doing maintenance on my hardware accelerators by
  using control-plane APIs to help coordinate maintenance and drain a Node.
- To support the new features, node maintenance, scheduler, descheduler, pod autoscaling, kubelet,
  and other actors should use a new eviction API to gracefully remove pods. This would enable new
  migration strategies that prefer to surge (upscale) pods first rather than downscale them. It
  would also allow other users/components to monitor pods that are gracefully removed/terminated
  and provide better behaviour in terms of de/scheduling, scaling and availability.
- As an end user, I would like more alternatives to blue-green upgrades, especially with special

>   hardware accelerators; it's far too expensive. I would like to choose a strategy on how to

Reviewer 13: ```suggestion
  hardware accelerators. I would like to choose a strategy on how to
```
That part of the statement doesn't provide any value to the use-case and it's just stating a fact, let's drop this.

  coordinate the node drain and the upgrade to achieve better cost-effectiveness.
- As a cloud provider, I need to perform regular maintenance on the hardware in my fleet. Enhancing

>   Kubernetes to help CSPs safely remove hardware will reduce operational costs.

Reviewer 13: What's a CSP? Can we introduce a abbreviation before using it, please?

> - As an end user or admin, I would like to use a mixture of on-demand and temporary spot instances

Reviewer 13: ```suggestion
- As an end user or admin, I would like to use a mixture of on-demand and spot instances
```
or 
```suggestion
- As an end user or admin, I would like to use a mixture of on-demand and temporary instances
```
No need to repeat that same thing twice

Author: +1

  in my clusters to reduce cloud expenditure. Having more reliable lifecycle and drain mechanisms
  for nodes will improve cluster stability in scenarios where instances may be terminated by the
  cloud provider due to cost-related thresholds.
- As a user, I want to prevent any disruption to my pet or expensive workloads (VMs, ML with
  accelerators) and either prevent termination for a long period of time or have a reliable
  migration path. Features like `terminationGracePeriodSeconds` are not sufficient as the
  termination/migration can take hours if not days.
- As a user, I want my application to finish all network and storage operations before terminating a
  pod. This includes closing pod connections, removing pods from endpoints, writing cached writes
  to the underlying storage and completing storage cleanup routines.
- As a cluster admin, I want a node to be declared as fully drained after all volumes are unmounted
  from it.

## Deliverables

The WG will coordinate requirement gathering, design, implementation, and progressing through
graduation stages.

> The group will help coordinate existing KEPs graduation as well as exploring new APIs and scenarios.

Reviewer 13: Same comment as before, introduce acronym before using it. This helps newcomers better understand what the document is talking about.

Area we expect to explore:

- An API to express node drain/maintenance.

> - An API to solve the problems wrt the API-initiated Eviction API and PDBs.

Reviewer 13: ```suggestion
- An API to solve the problems with the API-initiated Eviction API and PDBs.
```

- An API/mechanism to gracefully terminate pods during a node shutdown.
- An API to deschedule pods that use DRA devices.
- An API to remove pods from endpoints before they terminate.
- An API to track the schedulability (e.g. readiness) and capabilities of the node.
- Introduce enhancements across multiple Kubernetes SIGs to add support and integration for the new
  APIs to solve wide range of issues.

> We expect to provide reference implementations of the new APIs including but not limited to

Reviewer 6: this reference implementation will run in the kube-controller-manager, right? otherwise it will not be possible to run conformance

Author: Yes, part of KCM. Added a note.

controllers, API validation, integration with existing core components and extension points for the

> ecosystem. This should be accompanied by E2E / Conformance tests.

Reviewer 6: I still feel that an API that has to be so dependant on the infrastructure can be implemented in core in a general way, have you explored following the same pattern that Gateway API https://gateway-api.sigs.k8s.io/implementations/ followed , consolidating the APIs in official CRDs but leaving room for multiple implementations?

Author: The current plan is to bring in core API + main/reference implementation. It is still not certain if we will allow external controllers/implementations to control this API. Or if we will just allow well defined interactions.

If we have to offer multiple implementations we will explore and try to take inspiration from the Gateway API. Thanks for raising this.

## Relevant Features, KEPs and Documents

- Declarative Node Maintenance: https://github.com/kubernetes/enhancements/issues/4212
- EvictionRequest API: https://github.com/kubernetes/enhancements/issues/4563
- Graceful Node Shutdown [Beta]: https://github.com/kubernetes/enhancements/issues/2000

> - DRA: device taints and tolerations [Alpha]: https://github.com/kubernetes/enhancements/issues/5055

Reviewer 13: Nit: drop the stage from the title, it's orthogonal to what's being described here.

Author: We had people asking for this, but it is just one click away and can change over time

> - Disrupted Pods should be removed from endpoints: https://docs.google.com/document/d/1t25jgO_-LRHhjRXf4KJ5xY_t8BZYdapv7MDAxVGY6R8

Author: We will consider this issue as part of the other issue we will be solving. But still a good idea to mention the readiness probes explicitely. Plus I have added additional user story to cover this. @aojea  please let me know if it is sufficient.

- Implicit tolerations https://github.com/kubernetes/enhancements/issues/5282; add a Filter plugin to ensure that non-GPU pods are not scheduled on GPU nodes: https://github.com/kubernetes-sigs/scheduler-plugins/pull/812
- Node Resource Hot Plug: https://github.com/kubernetes/enhancements/issues/3953

## Relevant Projects

This is a list of known projects that solve similar problems in the ecosystem or would benefit from
the efforts of this WG:

- https://github.com/aws/aws-node-termination-handler
- https://github.com/foriequal0/pod-graceful-drain
- https://github.com/jukie/karpenter-deprovision-controller
- https://github.com/kubereboot/kured
- https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler
- https://github.com/kubernetes-sigs/cluster-api/
- https://github.com/kubernetes-sigs/karpenter
- https://github.com/kubernetes-sigs/kubespray
- https://github.com/kubevirt/kubevirt
- https://github.com/medik8s/node-maintenance-operator
- https://github.com/Mellanox/maintenance-operator
- https://github.com/NVIDIA/pika
- https://github.com/openshift/machine-config-operator
- https://github.com/planetlabs/draino
- https://github.com/strimzi/drain-cleaner

> There are also internal custom solutions that companies use.

Reviewer 13: ```suggestion
There are also internal, custom solutions that companies use.
```

## Prioritization

The group activity will focus on the following features:

- Declarative Node Maintenance
- EvictionRequest API
- Graceful Node Shutdown

And have the following priorities which mostly apply to WG calls, but can also apply to the general
WG review/work/guidance capacity:
1. Urgent topics concerning the WG focus features during the KEP and code greeze periods.
2. Discussing issues within the scope of the WG.
3. Presentations within the scope of the WG.

> 4. The WG focus features outside the KEP and code freeze periods.

Reviewer 13: Drop this point, it mostly expresses the same thing what the next one, only the latter uses better wording. This one suggests, the group will work outside of general release timeline, which is not allowed.

Author: done

5. Other WG features or features within the scope of the WG. If a topic is suggested multiple times,
   try to prevent starvation.

## Roles and Organization Management

This WG adheres to the Roles and Organization Management outlined in [wg-governance]
and opts-in to updates and modifications to [wg-governance].

[wg-governance]: /committee-steering/governance/wg-governance.md

## Timelines and Disbanding

The working group will disband once the features and core APIs defined in the following
KEPs/Features have reached a stable state (GA) and ongoing maintenance ownership is established
within the relevant SIGs. We will review whether the working group should disband if appropriate
SIG ownership can't be reached.

We expect to focus on bringing the following features to a stable state (GA):
- Declarative Node Maintenance
- EvictionRequest API

> - Graceful Node Shutdown

Reviewer 13: This is not needed, you're repeating what you've stated before more than once.

> We will also reevaluate the exit criteria upon the disbandment.

Reviewer 13: This is odd statement to put here, I'm not quite sure what you're trying to say here. The exit criteria define the disbandment, and you've clearly defined them above. They are not strict and they will be debatable, but this statement is contradicting that. I'd suggest to entirely drop this.

Author: Ack, updated.
